docs/real-time-analytics/mainframe-offload/process.sql

```diff
@@ -0,0 +1,37 @@
+SET 'auto.offset.reset' = 'earliest';
+
+-- Create stream of transactions from the Kafka topic
+CREATE STREAM mq_transactions (
+  dep_account_no STRING,
+  dep_balance_dollars BIGINT,
+  dep_balance_cents BIGINT,
+  timestamp BIGINT
+) WITH (
+  KAFKA_TOPIC = 'mq_transactions',
+  VALUE_FORMAT = 'JSON',
+  PARTITIONS = 6
+);
+
+-- Normalize the data and calculate timestamp deltas
+CREATE STREAM mq_transactions_normalized WITH (KAFKA_TOPIC = 'mq_transactions_normalized')
+  AS SELECT
+    dep_account_no,
+    CAST(dep_balance_dollars AS DECIMAL(10,2)) + CAST(dep_balance_cents AS DECIMAL(10,2))/CAST(100 as DECIMAL(10,2)) as balance,
+    timestamp AS ts_stream,
+    UNIX_TIMESTAMP() AS ts_cache,
+    (UNIX_TIMESTAMP() - timestamp) AS ts_delta
+FROM mq_transactions
+EMIT CHANGES;
+
+-- Materialize the stream of transactions into a table,
+-- like a local cache, that reflects the latest for each dep_account_no
+CREATE TABLE mq_cache WITH (KAFKA_TOPIC = 'mq_cache')
+  AS SELECT
+    dep_account_no,
+    latest_by_offset(BALANCE) AS balance_latest,
```

Comment:
Using `latest_by_offset` is outdated. We should use the new `CREATE SOURCE TABLE` instead

Reply:
@mjsax can you please elaborate on this?

1. Is there documentation that explains that `CREATE SOURCE TABLE` is a replacement for `latest_by_offset`?  I couldn't find it.
2. Is your suggestion to do this:

```
CREATE TABLE mq_cache WITH (KAFKA_TOPIC = 'mq_cache')
  AS SELECT *
FROM mq_transactions_normalized
GROUP BY dep_account_no
PARTITION BY dep_account_no
EMIT CHANGES;
```

3. There are other recipes that use `latest_by_offset` and it's not documented as being out of date, so the impact is cross-recipe potentially (i.e., maybe we tackle this in a separate issue for all recipes)

Reply:
@mjsax I found a PR for this update: https://github.com/confluentinc/ksql/pull/8458/files

Reply:
We are just about to update the docs... the PR you found is one piece -- we actually need a second PR for the docs.

The suggestion is to just do:
```sql
CREATE SOURCE TABLE mq_cache (
    dep_account_no STRING PRIMARY KEY,
    [...}  
) WITH(
    KAFKA_TOPIC='mq_transactions_normalized',
    KEY_FORMAT='KAFKA',
    VALUE_FORMAT='JSON')
```

A `CREATE SOURCE TABLE` has similar syntax than `CT` -- ie, it needs a scheme definition and you specify the input topic. -- The difference to `CT` is, that it's not just a metadata operation, but it deploy a KS program in the background that read the topic into RocksDB and thus the table is pull-query-able.

You can still materialize from an existing STREAM if you want, you you don't get proper UPSERT semantics -- an aggregation is not the same operation semantically -- in particular, there are not deletes.

Leave it up to your own judgment which pattern you want to use. Personally, I don't like `latest_by_offset` as it's quite broken semantically (and we only added it as cruel workaround -- I would always have preferred to fix it properly and not add `latest_by_offset` at all...)

Reply:
@mjsax I’m trying to rewrite this as suggested.

The first statement works:

```
CREATE SOURCE TABLE mq_transactions (
  dep_account_no STRING PRIMARY KEY,
  dep_balance_dollars BIGINT,
  dep_balance_cents BIGINT,
  timestamp BIGINT
) WITH (
  KAFKA_TOPIC = 'mq_transactions',
  VALUE_FORMAT = 'JSON',
  PARTITIONS = 6
);
```

But the manual INSERT INTO fails:

```
Cannot insert values into read-only table: MQ_TRANSACTIONS
```

From the docs: https://docs.ksqldb.io/en/latest/developer-guide/ksqldb-reference/create-table/

> Also, when you provide the SOURCE clause, the table is created as read-only. For a read-only table, INSERT statements and the DELETE TOPIC on DROP TABLE statements aren't be permitted.

Can you please confirm that I'm trying this as you intended, and that this is a real limitation that prevents us using this in the recipe?

Reply:
Rereading what you wrote

```
CREATE SOURCE TABLE mq_cache (
    dep_account_no STRING PRIMARY KEY,
    [...]
) WITH(
    KAFKA_TOPIC='mq_transactions_normalized',
    KEY_FORMAT='KAFKA',
    VALUE_FORMAT='JSON')
```

Maybe you intended something different, still using `mq_transactions_normalized`...stay tuned, I will try another set of commands and report back here

Reply:
Yes, the first queries should stay the same: we still process a STREAM (and we also want to INSERT VALUES into the STREAM).

Only the last step would be changed: instead of using an aggregation query on the STREAM, we create a SOURCE TABLE from the STREAM's topic.

Reply:
@mjsax thank you for all your help.  Queries now work, updated in https://github.com/confluentinc/ksqldb-recipes/pull/144/commits/a051ebdc1785bf8344f30161b7d9bd6689b4a25a